### PR TITLE
Convert synchronous funcs to async

### PR DESCRIPTION
**Pull request by kylecarbs**
**What this PR does / why we need it**:
Converts the extension to be entirely asynchronous. There were many unhandled asychronous functions that contained synchronous calls, which allowed tests to pass. Continuing with this could cause issues in the future.

**Which issue(s) this PR fixes**
None that I've found.

**Special notes for your reviewer**:
I'm not super familiar with this codebase so I'd be happy to hear feedback. I don't really enjoy the `load` function added to `historyFile.ts`.

### CODE REVIEW

**Comment by jpoon** [comment]
Some of these places make sense, but I think we need to be a bit more tactical about where we `await` for things as it has perf implications.

**Comment by kylecarbs** [comment]
All these changes await for funcs that were handled through executed some synchronous action (like fs.existsSync), so it should have similar perf as it was prior (its possible an async func slipped through).

I'd be happy to hear your suggestions.

**Comment by kylecarbs** [comment]
Any updates on this?

File: src/actions/commands/actions.ts [hunk 1]
```diff
@@ -140,7 +140,7 @@ export class DocumentContentChangeAction extends BaseAction {
         .add(last.positionDiff);
     }
 
-    vimState.currentMode = ModeName.Insert;
+    await vimState.setCurrentMode(ModeName.Insert);
```
Comment (kylecarbs) [inline src/actions/commands/actions.ts]: T

File: src/actions/commands/actions.ts [hunk 2]
```diff
@@ -2367,7 +2367,7 @@ class CommandExitVisualMode extends BaseCommand {
   keys = ['v'];
 
   public async exec(position: Position, vimState: VimState): Promise<VimState> {
-    vimState.currentMode = ModeName.Normal;
+    await vimState.setCurrentMode(ModeName.Normal);
```
Comment (kylecarbs) [inline src/actions/commands/actions.ts]: LGTM, will this have any perf implications?

File: src/util/historyFile.ts [hunk 3]
```diff
@@ -59,21 +59,23 @@ export class HistoryFile {
     }
   }
 
-  public async save(): Promise<void> {
-    return new Promise<void>((resolve, reject) => {
+  public save(): Promise<void> {
+    return new Promise<void>(async (resolve, reject) => {
```
Comment (jpoon) [inline src/util/historyFile.ts]: We can get rid of this `Promise` wrapper.

File: src/mode/modeHandlerMap.ts [hunk 4]
```diff
@@ -8,7 +8,16 @@ class ModeHandlerMapImpl {
     let modeHandler = this.modeHandlerMap[key];
     if (!modeHandler) {
       isNew = true;
-      modeHandler = await new ModeHandler();
+      await new Promise((res, rej) => {
```
Comment (jpoon) [inline src/mode/modeHandlerMap.ts]: Why was this necessary?

Comment (kylecarbs) [reply src/mode/modeHandlerMap.ts]: `ModeHandler` had async calls in its constructor that were not handled. I added the callback as an _initializer_ of sorts. I'm not a big fan of this method. We could add a static function and make the constructor private?
